cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/DataframeCollection.java

```diff
@@ -0,0 +1,526 @@
+import scala.collection.JavaConversions;
+import scala.collection.Seq;
+
+/**
+ * This SpartCollection stores data in Spark Dataframe ({@link Dataset}&lt;{@link Row}&gt;).
+ * From the {@link BatchCollection} API standpoint, it provides a collection of
+ * {@link StructuredRecord}, so any time a RDD or regular Dataset is requested, conversion
+ * to StructuredRecord will happen. This collection is useful to perform Spark SQL operations,
+ * especially chained ones, without unnessesary Dataset - RDD and StructuredRecord - Row
+ * conversions. Currently it supports Joiner and SparkCompute plugins.
+ */
+public class DataframeCollection extends DatasetCollection<StructuredRecord>
+    implements BatchCollection<StructuredRecord>{
+  private static final Logger LOG = LoggerFactory.getLogger(DataframeCollection.class);
+
+  private final Schema schema;
+  private final Dataset<Row> dataframe;
+  private Dataset<StructuredRecord> dataset;
+  public DataframeCollection(Schema schema, Dataset<Row> dataframe, JavaSparkExecutionContext sec,
+      JavaSparkContext jsc, SQLContext sqlContext, DatasetContext datasetContext,
+      SparkBatchSinkFactory sinkFactory, FunctionCache.Factory functionCacheFactory) {
+    super(sec, jsc, sqlContext, datasetContext, sinkFactory, functionCacheFactory);
+    this.schema = Objects.requireNonNull(schema);
+    this.dataframe = dataframe;
+  }
+
+  /**
+   * Allows to retrieve an underlying Dataframe.
+   *
+   * @return dataframe
+   */
+  public Dataset<Row> getDataframe() {
+    return dataframe;
+  }
+
+  @Override
+  protected synchronized Dataset<StructuredRecord> getDataset() {
+    if (dataset == null) {
+      //Copy schema to ensure we don't pull containing object into the lambda
+      Schema schema = this.schema;
+      MapFunction<Row, StructuredRecord> rowFunction = r -> DataFrames.fromRow(r, schema);
+      dataset = dataframe.map(rowFunction, objectEncoder());
+    }
+    return dataset;
+  }
+
+  @Override
+  public DataframeCollection toDataframeCollection(Schema schema) {
+    return this;
+  }
+
+  @Override
+  protected DataframeCollection cache(StorageLevel cacheStorageLevel) {
+    return wrap(dataframe.persist(cacheStorageLevel), schema);
+  }
+
+  @Override
+  public SparkCollection<StructuredRecord> join(JoinRequest joinRequest) {
+    Map<String, Dataset> collections = new HashMap<>();
+    String stageName = joinRequest.getStageName();
+    MapFunction<Row, Row> recordsInCounter = new CountingFunction<>(
+        stageName, sec.getMetrics(), Constants.Metrics.RECORDS_IN, sec.getDataTracer(stageName));
+    StructType leftSparkSchema = DataFrames.toDataType(joinRequest.getLeftSchema());
+    Dataset<Row> left = getDataframe().map(recordsInCounter, getDataframe().encoder());
+    collections.put(joinRequest.getLeftStage(), left);
+
+    List<Column> leftJoinColumns = joinRequest.getLeftKey().stream()
+        .map(left::col)
+        .collect(Collectors.toList());
+
+    /*
```

Comment:
this seems like it's just code that's been moved, please comment if there are any logical changes in this class that need a closer look.

Reply:
Yep, the Joins code was moved to DataframCollection and aggregation plugins code was moved to OpaqueDatasetCollection with no changes